defaultNatsService.go

```diff
@@ -0,0 +1,346 @@
+		authOpt := nats.UserCredentials(serviceName + ".creds")
+		return authOpt
+	})
+}
+
+func getDefaultService(lc fx.Lifecycle, config configuration.Config, logger log.Logger, getAuthOpts func(serviceName string) nats.Option) NatsService {
+	service := defaultNatsService{
+		logger:        logger,
+		subscriptions: map[string]*nats.Subscription{},
+	}
+	serviceName, err := config.GetServiceName()
+	if err != nil {
+		panic("can't get serviceName: " + err.Error())
+	}
+	natsUrl, err := config.Get("NatsUrl").String()
+	if err != nil {
+		panic("could not get conf by key NatsUrl " + err.Error())
+	}
+	connectionAttempts, err := config.Get("connectionAttempts").Int()
+	if err != nil {
+		connectionAttempts = 1
+	}
+	reconnectionAttempts, err := config.Get("reconnectionAttempts").Int()
+	if err != nil {
+		reconnectionAttempts = -1
+	}
+	authOpt := getAuthOpts(serviceName)
+	lc.Append(fx.Hook{
+		OnStart: func(c context.Context) error {
+			return service.Connect(natsUrl, serviceName, authOpt, connectionAttempts, reconnectionAttempts)
+		},
+		OnStop: func(c context.Context) error {
+			service.Close()
+			return nil
+		},
+	})
+	return &service
+}
+
+type defaultNatsService struct {
+	nc            NatsConnection
+	logger        log.Logger
+	subscriptions map[string]*nats.Subscription
+}
+
+func (n defaultNatsService) Connect(natsUrl, serviceName string, authOpt nats.Option, connectionAttempts, reconnectionAttempts int) error {
+	opts := []nats.Option{nats.Name(serviceName)}
+	opts = append(opts, authOpt)
+	opts = append(opts, nats.ReconnectWait(2*time.Second))
+	opts = append(opts, nats.MaxReconnects(reconnectionAttempts))
+	opts = append(opts, nats.MaxPingsOutstanding(2))
+	opts = append(opts, nats.PingInterval(2*time.Second))
+
+	opts = append(opts, nats.ErrorHandler(func(_ *nats.Conn, s *nats.Subscription, err error) {
+		n.logger.Error(context.Background(), "nats.ErrorHandler fired, Subscription: %v, err: %v", s, err)
+	}))
+
+	opts = append(opts, nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
+		n.logger.Error(context.Background(), "nats.DisconnectErrHandler fired, err: %v", err)
+	}))
+
+	opts = append(opts, nats.ReconnectHandler(func(nc *nats.Conn) {
+		n.logger.Warn(context.Background(), fmt.Sprintf("nats.ReconnectHandler [%s]", nc.ConnectedUrl()))
+	}))
+
+	err := n.connect(natsUrl, opts)
+	if err != nil {
+		n.logger.Error(context.Background(), "can't connect to nats server %s err: %v", natsUrl, err)
+		if connectionAttempts < 2 {
+			return err
+		}
+		go func() {
+			for i := 1; i <= connectionAttempts; i++ {
+				time.Sleep(time.Minute)
+				err := n.connect(natsUrl, opts)
+				if err == nil {
+					break
+				}
+				n.logger.Error(context.Background(), "can't connect to nats server %s err: %v. attempt - %v", natsUrl, err, i)
+			}
+			n.logger.Error(context.Background(), "can't connect to nats server %s after %v attempts. Stop trying", natsUrl, connectionAttempts)
```

Comment:
and...now I have a service that will NEVER connect ?
we should probably have a hook here, so the consumer can register and decide what to do (panic?)